### ISSUE TITLE
Heredoc strings no longer syntax highlighted

### ISSUE THREAD

**Issue by andremleblanc** [atom/language-ruby#249]
* [ ] Put an X between the brackets on this line if you have done all of the following:
    * Reproduced the problem in Safe Mode: http://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode
    * Followed all applicable steps in the debugging guide: http://flight-manual.atom.io/hacking-atom/sections/debugging/
    * Checked the FAQs on the message board for common solutions: https://discuss.atom.io/c/faq
    * Checked that your issue isn't already filed: https://github.com/issues?utf8=✓&q=is%3Aissue+user%3Aatom
    * Checked that there is not already an Atom package that provides the described functionality: https://atom.io/packages

### Description

Heredocs used to use the syntax highlighting for that language, but it appears that isn't currently working.

Before:
<img width="202" alt="screen shot 2018-11-02 at 10 25 07 pm" src="https://user-images.githubusercontent.com/5190468/47947208-6ef60600-deee-11e8-965a-df4db6c04b88.png">

Current:
<img width="200" alt="screen shot 2018-11-02 at 10 24 53 pm" src="https://user-images.githubusercontent.com/5190468/47947204-6998bb80-deee-11e8-8d3c-3825dc0ca90d.png">

### Steps to Reproduce

Use heredoc syntax in Ruby grammar.

**Expected behavior:**

It highlights the content of the heredoc in that language's syntax.

**Actual behavior:**

It doesn't.

**Reproduces how often:**

100%

### Versions

1.32.1

**Comment by JoshCheek** [comment]
I'm hitting both issues:

* Syntax highlighting doesn't happen in a heredoc anymore
* I can reproduce @Michael-Ludgate's issue, too (it's like it doesn't realize it's in the string), but only when I use certain delimiters. Eg it doesn't happen if I use a delimiter of `HERE`, or `RUBY`, but it does happen if I use a delimiter of `HEREDOC` or `ABCD`.

Anyway, the syntax highlighting is a huge feature for me, depending on what I'm doing, I might go to Atom just b/c of this feature (eg when I'm playing with SQL).

**Comment by JoshCheek** [comment]
Oh, that second one is interesting, it seems to be related to the letter `C`.

Put this into your editor, select the `ABCD` and do `Cmd+d` to highlight the second one. Then slowly type `ABCD`, it's highlighted as a string for `AB`, but then loses the highlighting on the `C`. If you skip `C` and go to `D`, it remains highlighted correctly. We can see the same thing in @Michael-Ludgate's example, the word `HEREDOC` has a `C` in it, and it correctly highlights as a string up until that letter is typed. If you change it to `HEREDOK`, it doesn't get messed up.

```ruby
def m
  <<~ABCD
  msg
  ABCD
end
```

**Comment by JoshCheek** [comment]
Oh, okay, so it's going wonky on the highlighting because it applies "C" language syntax highlighting to it. So syntax highlighting does still get applied, but only sometimes. Maybe there's an issue with how it looks up the language? Eg here it sees `HEREDOC`, notices the `C` in there, and applies C syntax highlighting, do `SQL`, and it doesn't recognize it as SQL. Inspecting the DOM shows it's been given a class of `syntax--string`.

![image](https://user-images.githubusercontent.com/77495/50525307-46c6f580-0aa0-11e9-880a-1076859da473.png)

**Comment by JoshCheek** [comment]
Followed the instructions to make changes locally ([here](https://flight-manual.atom.io/hacking-atom/sections/hacking-on-atom-core/) to put atom into dev mode, and [here](https://flight-manual.atom.io/hacking-atom/sections/contributing-to-official-atom-packages/) to link local changes into the dev-mode). Ran a `git-bisect` on it.

[This](701ad5d744e23090655f1c2e725f18cfe6f4ae7c) is the commit where it breaks.

Here's the output of the `git bisect` (in another window I was restarting atom each time I did this and looking to see if it would highlight SQL correctly):

```sh
🐠  git bisect start

🐠  git bisect bad

🐠  git bisect good 0909a11557725f117dea6f39aa2b08241a887ff2
Bisecting: 31 revisions left to test after this (roughly 5 steps)
[726281fc9d9fe4e9e3e20a61f5912b062625edcb] Merge pull request #236 from atom/mb-compatible-scope-descriptors

🐠  git bisect bad
Bisecting: 14 revisions left to test after this (roughly 4 steps)
[e5a8d8e66728c013e1a009bcdb2facfd044270f9] Merge pull request #225 from atom/mb-tree-sitter-parser

🐠  git bisect bad
Bisecting: 7 revisions left to test after this (roughly 3 steps)
[701ad5d744e23090655f1c2e725f18cfe6f4ae7c] Add tree-sitter Ruby grammar

🐠  git bisect bad
Bisecting: 4 revisions left to test after this (roughly 2 steps)
[73ec57162fb7eb0035dedc0026911e3d2a8aee4b] :racehorse: Improve the regexp pattern

🐠  git bisect good
Bisecting: 2 revisions left to test after this (roughly 1 step)
[ef4eb5691842a369b83f6465defaad8b82ad191a] Merge pull request #215 from stevenpetryk/variables-ending-in-do-are-not-blocks

🐠  git bisect good
Bisecting: 0 revisions left to test after this (roughly 1 step)
[d88a4cfb32876295ec5b090a2994957e62130f4a] Prepare 0.71.4 release

🐠  git bisect good
701ad5d744e23090655f1c2e725f18cfe6f4ae7c is the first bad commit
commit 701ad5d744e23090655f1c2e725f18cfe6f4ae7c
Author: Max Brunsfeld <maxbrunsfeld@gmail.com>
Date:   Fri Dec 8 09:28:08 2017 -0800

    Add tree-sitter Ruby grammar

:040000 040000 32ce74581b8f1777099d63e11f36f2625814dd82 adcea68aa1e140b88926e33c0f3849582360dce9 M	grammars
:100644 100644 0db247528eaa512b24d4a033c36835c164453f75 c88b06b4fee6a4c54319b0c8b1ae1107037a249a M	package.json
```

-----

Additionally, I put print statements into the injection point language registration (line 6):

https://github.com/atom/language-ruby/blob/ea3824d2539fe982e69d0599c5fadd082879f1c6/lib/main.js#L4-L8

For `<<~HEREDOC`, it was returning the language of `"HEREDOC"`, but for `"SQL"`, it was returning an empty string (or possibly something that prints like an empty string).

**Comment by JoshCheek** [comment]
There are apparently multiple grammars which are loaded, and the one that is used is determined by a configuration value. This package uses this feature to test against the old parser (textmate's?), which means that the tree-sitter stuff doesn't get tested by this lib.

On the bright side, it reveals an easy solution: 

```js
atom.config.set('core.useTreeSitterParsers', false)
```

If you enter it into the debug console (`cmd-opt-i`) it will immediately update to be correct, and will persist this change in Atom's `config.cson`.

![image](https://user-images.githubusercontent.com/77495/50543046-80772980-0b90-11e9-9f0c-72bdd62fd01d.png)

**Comment by BerkhanBerkdemir** [comment]
I can reproduce this issue with Atom 1.35.1 and Ubuntu 18.04; however, https://github.com/atom/language-ruby/issues/249#issuecomment-450528828 is useful.

**Comment by jasonrudolph** [comment]
Thanks for reporting this issue. I think I see what's causing the behavior that you're observing. I'll try to explain what I've found so far and a path toward resolving this issue.

**tl;dr**: When using Tree-sitter, the Ruby heredoc's inner text will be syntax highlighted if (1) there is Tree-sitter grammar for the language of the inner text and (2) that grammar provides an injection regex. Currently, JavaScript, Bash, C, and a [handful of other languages](https://github.com/atom/flight-manual.atom.io/blob/a249e69d64d8a622fd704affbe4f0b74b7269bb4/content/hacking-atom/sections/creating-a-grammar.md#example-packages) meet these requirements, but SQL does not yet meet these requirements.

---

## Tree-sitter support for syntax highlighting within Ruby heredocs
Tree-sitter refers to this topic as "[language injection](https://github.com/atom/flight-manual.atom.io/blob/a249e69d64d8a622fd704affbe4f0b74b7269bb4/content/hacking-atom/sections/creating-a-grammar.md#language-injection)" and describes it as follows:

> Sometimes, a source file can contain code written in several different languages. Tree-sitter grammars support this situation using a two-part process called *language injection*. First, an 'outer' language must define an *injection point* - a set of syntax nodes whose text can be parsed using a different language, along with some logic for guessing the *name* of the other language that should be used. Second, an 'inner' language must define an `injectionRegex` - a regex pattern that will be tested against the language name provided by the injection point.

The Tree-sitter Ruby grammar satisfies the first requirement described above: the grammar defines an [injection point for heredocs](https://github.com/atom/language-ruby/blob/0592946a72481f307860927e8c5a37c705388985/lib/main.js#L4-L8). 😅

## Some languages already support syntax highlighting within Ruby heredocs
Several languages satisfy the second requirement described above: they define an `injectionRegex`. For example, language-javascript defines one [here](https://github.com/atom/language-javascript/blob/6e90f7bd95b7a13e4a80834ce500d9a952b25ba8/grammars/tree-sitter-javascript.cson#L8), and language-shellscript defines one [here](https://github.com/atom/language-shellscript/blob/904a01c0895740776845d4e915d2399c5e4d8c6c/grammars/tree-sitter-bash.cson#L11). We can see that Atom provides syntax highlighting for these languages within a Ruby heredoc:

<img width="356" alt="Screen Shot 2019-04-24 at 4 23 38 PM" src="https://user-images.githubusercontent.com/2988/56692595-8b5e4880-66b0-11e9-9838-7eb2ae020029.png">

<img width="344" alt="Screen Shot 2019-04-24 at 4 25 26 PM" src="https://user-images.githubusercontent.com/2988/56692596-8bf6df00-66b0-11e9-9d4c-7f733201b0d5.png">

## SQL doesn't yet support syntax highlighting within Ruby heredocs
At the moment, [language-sql](https://github.com/atom/language-sql) does not yet provide a Tree-sitter grammar, so Atom cannot provide SQL syntax highlighting inside a Ruby heredoc:

<img width="363" alt="Screen Shot 2019-04-24 at 4 30 30 PM" src="https://user-images.githubusercontent.com/2988/56692598-8bf6df00-66b0-11e9-9e35-1f3020a78e7e.png">

## Fixing this issue
To fix this issue, we'll need to update Atom's other `language-*` packages to (1) provide a Tree-sitter grammar and (2) provide an `injectionRegex`.


## A workaround in the meantime
The fix above will be a rather time consuming endeavor, so folks may want a workaround in the meantime.  As @JoshCheek notes [above](https://github.com/atom/language-ruby/issues/249#issuecomment-450528828), you can disable Tree-sitter for now. To do so, open Atom's [settings UI](https://flight-manual.atom.io/getting-started/sections/atom-basics/#settings-and-preferences), and uncheck the "Use Tree Sitter Parsers" checkbox. 

<img width="1034" alt="Screen Shot 2019-04-24 at 4 39 54 PM" src="https://user-images.githubusercontent.com/2988/56692599-8bf6df00-66b0-11e9-8d3d-3166ac98f0cd.png">

This will cause Atom to fall back to using the TextMate grammars, which currently provide broader support for syntax highlighting within Ruby heredocs.

**Comment by BerkhanBerkdemir** [comment]
**TL;DR** there is nothing wrong with `language-ruby`, and this is because of `language-sql`. Am I right :laughing:

**Comment by CodingMarkus** [comment]
> Put this into your editor, select the `ABCD` and do `Cmd+d` to highlight the second one. Then slowly type `ABCD`, it's highlighted as a string for `AB`, but then loses the highlighting on the `C`. If you skip `C` and go to `D`, it remains highlighted correctly. We can see the same thing in @Michael-Ludgate's example, the word `HEREDOC` has a `C` in it, and it correctly highlights as a string up until that letter is typed. If you change it to `HEREDOK`, it doesn't get messed up.

Isn't that already a bug on its own or should really any HEREDOC that contains a capital C cause the content to be interpreted as C code? Sounds a bit like a `=~ /C/` is done when a `== "C"` would have been correct.

**Comment by nathansobo** [comment]
Another possibly resolution would be to support injecting legacy first-mate grammars into regions of files that are otherwise parsed via TreeSitter. This would need to occur in Atom rather than TreeSitter, and I'm not sure how complicated it would be.

**Comment by JoshCheek** [comment]
Maybe worth noting that "Github Markdown" highlights it correctly (LHS), but Ruby doesn't.

![image](https://user-images.githubusercontent.com/77495/108024351-01e8da80-6fea-11eb-9a1d-6cbaedc5333a.png)

The scopes seem really different, almost seems like they don't use the same Ruby parser.

![image](https://user-images.githubusercontent.com/77495/108024730-b84cbf80-6fea-11eb-9692-2e06ba4f2e03.png)

**Comment by lacostenycoder** [comment]
Is this a dead issue? I remember this working at some point...
